From 5b571ca7c77c4ff1260d7c0347bbd885c3589cd0 Mon Sep 17 00:00:00 2001
From: Matthew Ballance <matt.ballance@gmail.com>
Date: Sun, 1 Feb 2026 22:24:24 +0000
Subject: [PATCH] build: add device SDK installation

Install device development SDK headers and pkg-config file to enable
building QEMU device modules as external shared libraries.

The SDK includes:
- Public API headers from include/{qemu,qom,hw,exec,sysemu,chardev,io,migration,monitor}
- Generated configuration files (config-host.h, config-poison.h)
- Generated QAPI headers
- pkg-config file for easy integration

Headers are installed to ${prefix}/include/qemu-device/
---
 meson.build | 64 +++++++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 64 insertions(+)

diff --git a/meson.build b/meson.build
index 147097c..f047f50 100644
--- a/meson.build
+++ b/meson.build
@@ -4320,6 +4320,70 @@ if get_option('plugins')
     # so that plugin authors can compile against it.
     install_data(win32_qemu_plugin_api_lib, install_dir: 'lib')
   endif
+
+###########################
+# Device Module SDK       #
+###########################
+
+# Install all public API headers needed for device development
+install_subdir('include/qemu',
+               install_dir: get_option('includedir') / 'qemu-device',
+               strip_directory: false,
+               exclude_files: ['osdep.h.inc'])
+
+install_subdir('include/qom',
+               install_dir: get_option('includedir') / 'qemu-device',
+               strip_directory: false)
+
+install_subdir('include/hw',
+               install_dir: get_option('includedir') / 'qemu-device',
+               strip_directory: false)
+
+install_subdir('include/exec',
+               install_dir: get_option('includedir') / 'qemu-device',
+               strip_directory: false)
+
+install_subdir('include/sysemu',
+               install_dir: get_option('includedir') / 'qemu-device',
+               strip_directory: false)
+
+install_subdir('include/chardev',
+               install_dir: get_option('includedir') / 'qemu-device',
+               strip_directory: false)
+
+install_subdir('include/io',
+               install_dir: get_option('includedir') / 'qemu-device',
+               strip_directory: false)
+
+install_subdir('include/migration',
+               install_dir: get_option('includedir') / 'qemu-device',
+               strip_directory: false)
+
+install_subdir('include/monitor',
+               install_dir: get_option('includedir') / 'qemu-device',
+               strip_directory: false)
+
+# Install generated configuration files
+install_data(meson.current_build_dir() / 'config-poison.h',
+             install_dir: get_option('includedir') / 'qemu-device')
+install_data(meson.current_build_dir() / 'config-host.h',
+             install_dir: get_option('includedir') / 'qemu-device')
+
+# Install generated QAPI headers (subdirectory with all files)
+install_subdir(meson.current_build_dir() / 'qapi',
+               install_dir: get_option('includedir') / 'qemu-device',
+               strip_directory: false,
+               exclude_directories: ['trace'])
+
+# Generate and install pkg-config file
+pkg = import('pkgconfig')
+pkg.generate(
+  name: 'qemu-device',
+  description: 'QEMU Device Development SDK',
+  version: meson.project_version(),
+  subdirs: ['qemu-device'],
+)
+
 endif
 
 subdir('qga')
-- 
2.43.0

